### PR TITLE
TLS: Fix issuerDN for created certificates

### LINKED ISSUE #69403: CertGenUtils sets issuer DN incorrectly
https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/CertGenUtils.java#L179

When generating a new certificate we use the _issuer_ of the CA rather than the _subject_ of the CA.
For self-signed CAs (which is the only type of CA that certutil can generate), this works fine, but for intermediate CAs, it is incorrect.

Reported: #53584

### PR DESCRIPTION
**Pull request by justincr-elastic**
Certificate chains require subjectCert.issuerDN = issuerCert.subjectDN.
Each adjacent certificate must satisfy this relationship, or else they
fail certificate chain validation.

CertGenUtils was setting subjectCert.issuerDN = issuerCert.issuerDN.
That was incorrect. If the issuerCert was for a root CA, then issuerCert
had the same value for issuerCert.subjectDN = issuerCert.issuerDN, and
it was not possible to detect the bug.

For testing, use a subordinate CA to create a TLS certificate. A
subordinate CA has different values for its issuerCert.subjectDN !=
issuerCert.issuerDN. When the TLS certificate is created, it becomes
possible to assert subjectCert.issuerDN = issuerCert.subjectDN for the
TLS certificate.

Closes #69403

### CODE REVIEW

**Comment by justincr-elastic** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java [hunk 1]
```diff
@@ -100,4 +105,31 @@ public void testIsAnyLocalAddress() throws Exception {
         verifyNoMoreInteractions(address);
     }
 
+    public void testIssuerCertSubjectDN() throws Exception {
+        final ZonedDateTime notBefore = LocalDateTime.now().atZone(ZoneId.of("UTC"));
```
Comment (tvernum) [inline x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:109]: This line is failing the forbidden APIs check. You probably want:
```suggestion
        final ZonedDateTime notBefore = ZonedDateTime.now(ZoneOffset.UTC);
```

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java [hunk 2]
```diff
@@ -100,4 +105,31 @@ public void testIsAnyLocalAddress() throws Exception {
...
+        // end entity
+        final X500Principal endEntityPrincipal = new X500Principal("CN=TLS Client\\+Server,DC=Sub CA,DC=example.com");
+        final KeyPair endEntityKeyPair = CertGenUtils.generateKeyPair(2048);
+        final X509Certificate endEntityCert = CertGenUtils.generateSignedCertificate(endEntityPrincipal, null, endEntityKeyPair,
+            subCaCert, subCaKeyPair.getPrivate(), true, notBefore, notAfter, null);
+
+        // verify generateSignedCertificate performed DN chaining correctly
+        assertEquals(subCaCert.getSubjectX500Principal(), endEntityCert.getIssuerX500Principal());
+        assertEquals(rootCaCert.getSubjectX500Principal(), subCaCert.getIssuerX500Principal());
+        assertEquals(rootCaCert.getSubjectX500Principal(), rootCaCert.getIssuerX500Principal());
```
Comment (tvernum) [inline x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:133]: We generally prefer to use Hamcrest style assertions (although we also have a lot of older code that uses the standard JUnit assertion style).
Matcher-based assertions give much better failure messages which has an enormous benefit over a large codebase like ES (the person triaging a failure has typically not seen the test before, and the better the assertion failure message is, the easier & quicker it is to resolve).

 ```suggestion
        assertThat(rootCaCert.getIssuerX500Principal(), Matchers.equalTo(rootCaCert.getSubjectX500Principal()));
```
(though you would also want to static import `Matchers.equalTo`, but :octocat: suggestions don't make that easy to show).

**Review by tvernum — COMMENTED** [review]
This looks like the right approach for both the change & the test.
I left a couple of comments about some ES specific style/API choices.

**Comment by justincr-elastic** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java (hunk 1 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java]: I updated the two dates as follows:

```
        final ZonedDateTime notBefore = ZonedDateTime.now(ZoneOffset.UTC);
        final ZonedDateTime notAfter = ZonedDateTime.parse("2099-12-31T23:23:59.999999+00:00");
```

They passes the forbidden API check:

`
./gradlew :x-pack:plugin:security:cli:forbiddenApisTest
`

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java (hunk 2 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java]: I switched to Matcher.equalsTo().

The updates pass the style check.
```
./gradlew :x-pack:plugin:security:cli:checkstyleTest
```

**Comment by justincr-elastic** [comment]
All checks have passed. Will merge if the update is approved.

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java [hunk 3]
```diff
@@ -100,4 +109,87 @@ public void testIsAnyLocalAddress() throws Exception {
...
+
+        // EKU=null|clientAuth, KU=digitalSignature, authType=DHE_DSS/DHE_RSA/ECDHE_ECDSA/ECDHE_RSA/RSA_EXPORT/UNKNOWN
+        if (doTlsClientCheck) {
+            x509ExtendedTrustManager.checkClientTrusted(certChain, "ECDHE_RSA");
+        }
+
+        // EKU=null|serverAuth, KU=digitalSignature, authType=DHE_DSS/DHE_RSA/ECDHE_ECDSA/ECDHE_RSA/RSA_EXPORT/UNKNOWN
+        // EKU=null|serverAuth, KU=keyEncipherment,  authType=RSA
+        // EKU=null|serverAuth, KU=keyAgreement,     authType=DH_DSS/DH_RSA/ECDH_ECDSA/ECDH_RSA
+        if (doTlsClientCheck) {
```
Comment (jkakavas) [inline x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:191]:
```suggestion
        if (doTlsServerCheck) {
```

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java [hunk 4]
```diff
@@ -100,4 +109,87 @@ public void testIsAnyLocalAddress() throws Exception {
...
+
+        // verify cert chaining based on PKIX rules (ex: SubjectDNs/IssuerDNs, SKIs/AKIs, BC, KU, EKU, etc)
+        final KeyStore trustStore = KeyStore.getInstance("PKCS12", "SunJSSE"); // EX: SunJSSE, BC, BC-FIPS
+        trustStore.load(null, null);
+        trustStore.setCertificateEntry("trustAnchor", rootCaCert); // anchor: any part of the chain, or issuer of last entry in chain
+
+        validateEndEntityTlsChain(trustStore, certChain, true, true);
+    }
+
+    // Wrap TrustStore in TrustManager[], cast first TrustManager to X509ExtendedTrustManager to call validation methods
```
Comment (jkakavas) [inline x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:149]: misplaced comment ?

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java [hunk 5]
```diff
@@ -100,4 +109,87 @@ public void testIsAnyLocalAddress() throws Exception {
...
+            }
+        }
+        if (null == x509ExtendedTrustManager) {
+            throw new UnsupportedOperationException("Expected at least one javax.net.ssl.X509ExtendedTrustManager");
+        }
+
+        // TLS authType is a substring of cipher suite. It controls what OpenJDK EndEntityChecker.java checks in KU/EKU/etc extensions.
+
+        // EKU=null|clientAuth, KU=digitalSignature, authType=DHE_DSS/DHE_RSA/ECDHE_ECDSA/ECDHE_RSA/RSA_EXPORT/UNKNOWN
+        if (doTlsClientCheck) {
```
Comment (jkakavas) [inline x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:184]: For my education mostly: What additional checks would `checkClientTrusted` do in this case compared to `checkServerTrusted` regarding the validity of the chain ?

**Review by jkakavas — COMMENTED** [review]
LGTM, I left one comment to be resolved before merging. Also left a nit and a question but happy for this to be merged when the comment is addressed

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java (hunk 5 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:184]: Both do same CA cert checks. Differences only apply to the first cert in the chain (i.e. the end entity cert).
- _Subject Alt Name_: Server hostname checking uses SAN dnsName & ipAddress. Client authN doesn't do anything. The caller is free to implement something else based on use case (ex: SubjectDN or SAN.directoryName for LDAP, SAN.otherName for Win DC legacy SAM account, SAN.rfc822Name for email address).
- _Key Usage_: Client authN cipher suites support RSA sign & EC sign. Server supports RSA sign, EC sign, RSA encrypt, and ECDH key agreement. Server has more flexibility.
- _Extended Key Usage_: Server supports null EKU or serverAuth. Client supports null EKU or clientAuth.

I learned by skimming through EndEntityChecker.check(). It is well documented and easy to understand. Drill down into checkTLSClient() and checkTLSServer(), and you can see the differences I mentioned for KU and EKU.
- https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/sun/security/validator/EndEntityChecker.java#L142

```
    void check(X509Certificate cert, Object parameter, boolean checkUnresolvedCritExts) throws CertificateException {
        if (variant.equals(Validator.VAR_GENERIC)) {
            return; // no checks
        }
        Set<String> exts = getCriticalExtensions(cert);
        if (variant.equals(Validator.VAR_TLS_SERVER)) {
            checkTLSServer(cert, (String)parameter, exts);                  // <===== Check this method
        } else if (variant.equals(Validator.VAR_TLS_CLIENT)) {
            checkTLSClient(cert, exts);                                     // <===== Check this method
        } else if (variant.equals(Validator.VAR_CODE_SIGNING)) {
            checkCodeSigning(cert, exts);
        } else if (variant.equals(Validator.VAR_JCE_SIGNING)) {
            checkCodeSigning(cert, exts);
        } else if (variant.equals(Validator.VAR_PLUGIN_CODE_SIGNING)) {
            checkCodeSigning(cert, exts);
        } else if (variant.equals(Validator.VAR_TSA_SERVER)) {
            checkTSAServer(cert, exts);
        } else {
            throw new CertificateException("Unknown variant: " + variant);
        }
        // if neither VAR_GENERIC variant nor unknown variant
        if (checkUnresolvedCritExts) {
            checkRemainingExtensions(exts);
        }
    }
```

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java (hunk 4 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:149]: Yes. I will remove it.

File: x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java (hunk 3 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java:191]: Good catch. This is a copy and paste mistake on my part. I will fix it.

**Review by jkakavas — APPROVED** [review]
LGTM

**Pull request merged**